packages/tizen_rpc_port/example/example_proxy/lib/message_proxy.dart

```diff
@@ -0,0 +1,262 @@
+  Future<void> onConnected();
+
+  /// This abstract method will be called when the connection with the stub is disconnected.
+  Future<void> onDisconnected();
+
+  /// This abstract method will be called when connect() is failed.
+  Future<void> onRejected(String errorMessage);
+
+  @override
+  @nonVirtual
+  Future<void> onConnectedEvent() async {
+    _online = true;
+    await onConnected();
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onDisconnectedEvent() async {
+    _online = false;
+    await onDisconnected();
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onRejectedEvent(String errorMessage) async {
+    await onRejected(errorMessage);
+  }
+
+  @override
+  @nonVirtual
+  Future<void> onReceivedEvent(Parcel parcel) async {
+    final int cmd = parcel.readInt32();
+    if (cmd != _MethodId.callback.id) {
+      return;
+    }
+
+    await _processReceivedEvent(parcel);
+  }
+
+  Future<void> _processReceivedEvent(Parcel parcel) async {
+    final int id = parcel.readInt32();
+    final int seqId = parcel.readInt32();
+    final bool once = parcel.readBool();
+
+    for (final CallbackBase delegate in _delegateList) {
+      if (delegate._id == id && delegate._seqId == seqId) {
+        await delegate._onReceivedEvent(parcel);
+        if (delegate._once && once) {
+          _delegateList.remove(delegate);
+        }
+        break;
+      }
+    }
+  }
+
+  Future<Parcel> _consumeCommand(Port port) async {
+    while (true) {
+      try {
+        final Parcel parcel = Parcel.fromPort(port);
+        final int cmd = parcel.readInt32();
+        if (cmd == _MethodId.result.id) {
+          return parcel;
+        }
+      } catch (e) {
+        Log.error(_logTag, e.toString());
+        return Parcel();
+      }
+    }
+  }
+
+  /// Connects with the stub application.
+  @override
+  Future<void> connect() async {
+    Log.info(_logTag, 'connect()');
+    await super.connect();
+  }
+
+  /// Dispose registered delegate interface.
+  void disposeCallback(String tag) {
+    Log.info(_logTag, 'disposeCallback($tag)');
+    _delegateList.removeWhere((CallbackBase element) => element.tag == tag);
+  }
+
+  /// This method is used to send 'Register' request to the stub app.
+  Future<int> register(String name, NotifyCB cb) async {
+    Log.info(_logTag, 'Register');
+
+    if (!_online) {
+      throw Exception('NotConnectedSocketException');
+    }
+
+    final Port port = getPort(PortType.main);
+
+    final Parcel parcel = Parcel();
+    final ParcelHeader header = parcel.header;
+    header.tag = _tidlVersion;
+    parcel.writeInt32(_MethodId.register.id);
+
+    parcel.writeString(name);
+    cb.serialize(parcel);
+    _delegateList.add(cb);
+
+    parcel.send(port);
+
+    late Parcel parcelReceived;
+    while (true) {
+      parcelReceived = await _consumeCommand(port);
+      final ParcelHeader headerReceived = parcelReceived.header;
+      if (headerReceived.tag.isEmpty) {
+        break;
```

Comment:
What does the empty tag mean?

Reply:
If stub interface file is old version, it is empty

Reply:
In that case shouldn't you just ignore the tag but just compare `sequenceNumber`s?

Reply:
```dart
    resultHeader.sequenceNumber = header.sequenceNumber;
```
`sequenceNumber` also need to set from stub.
So, if stub interface is old. it hard to compare it.

Reply:
The parcel header was added since Tizen 6.5. The RPC-Port API was provided since Tizen 4.0.
The parcel header is for checking the protocol validation.
If the stub or the proxy uses the generated code using the version of the TIDL generator lower than '1.5.0', the tag is always empty string.

Reply:
@hjhun Okay, then shouldn't the return type of `Parcel.header` be nullable? Does `rpc_port_parcel_get_header` return a null pointer if the header is not found?

Reply:
@swift-kim,
Even if the application was developed using tizen 4.0 base, the application can run on the tizen 6.5 or latest version.
Parcel.header cannot be nullable. The header of the parcel is always created since Tizen 6.5.

Even if the app developed in the previous version (e.g. tizen 4.0) does not use the Parcel header, the Parcel header is delivered to the app developed in the latest version. However, the tag is empty string.

Reply:
Okay, thank you for the explanation. It was not clear from the `rpc_port_parcel_get_header` documentation.